// es-abstract/operations/2015.js
'use strict';

module.exports = {
	IsPropertyDescriptor: 'https://262.ecma-international.org/6.0/#sec-property-descriptor-specification-type', // not actually an abstract op

	abs: {
		url: 'https://262.ecma-international.org/6.0/#sec-algorithm-conventions'
	},
	'Abstract Equality Comparison': {
		url: 'https://262.ecma-international.org/6.0/#sec-abstract-equality-comparison'
	},
	'Abstract Relational Comparison': {
		url: 'https://262.ecma-international.org/6.0/#sec-abstract-relational-comparison'
	},
	AddRestrictedFunctionProperties: {
		url: 'https://262.ecma-international.org/6.0/#sec-addrestrictedfunctionproperties'
	},
	AdvanceStringIndex: {
		url: 'https://262.ecma-international.org/6.0/#sec-advancestringindex'
	},
	AllocateArrayBuffer: {
		url: 'https://262.ecma-international.org/6.0/#sec-allocatearraybuffer'
	},
	AllocateTypedArray: {
		url: 'https://262.ecma-international.org/6.0/#sec-allocatetypedarray'
	},
	ArrayCreate: {
		url: 'https://262.ecma-international.org/6.0/#sec-arraycreate'
	},
	ArraySetLength: {
		url: 'https://262.ecma-international.org/6.0/#sec-arraysetlength'
	},
	ArraySpeciesCreate: {
		url: 'https://262.ecma-international.org/6.0/#sec-arrayspeciescreate'
	},
	BoundFunctionCreate: {
		url: 'https://262.ecma-international.org/6.0/#sec-boundfunctioncreate'
	},
	Call: {
		url: 'https://262.ecma-international.org/6.0/#sec-call'
	},
	Canonicalize: {
		url: 'https://262.ecma-international.org/6.0/#sec-runtime-semantics-canonicalize-ch'
	},
	CanonicalNumericIndexString: {
		url: 'https://262.ecma-international.org/6.0/#sec-canonicalnumericindexstring'
	},
	CharacterRange: {
		url: 'https://262.ecma-international.org/6.0/#sec-runtime-semantics-characterrange-abstract-operation'
	},
	CharacterSetMatcher: {
		url: 'https://262.ecma-international.org/6.0/#sec-runtime-semantics-charactersetmatcher-abstract-operation'
	},
	CloneArrayBuffer: {
		url: 'https://262.ecma-international.org/6.0/#sec-clonearraybuffer'
	},
	CompletePropertyDescriptor: {
		url: 'https://262.ecma-international.org/6.0/#sec-completepropertydescriptor'
	},
	Completion: {
		url: 'https://262.ecma-international.org/6.0/#sec-implicit-completion-values'
	},
	CompletionRecord: {
		url: 'https://262.ecma-international.org/6.0/#sec-completion-record-specification-type'
	},
	Construct: {
		url: 'https://262.ecma-international.org/6.0/#sec-construct'
	},
	CopyDataBlockBytes: {
		url: 'https://262.ecma-international.org/6.0/#sec-copydatablockbytes'
	},
	CreateArrayFromList: {
		url: 'https://262.ecma-international.org/6.0/#sec-createarrayfromlist'
	},
	CreateArrayIterator: {
		url: 'https://262.ecma-international.org/6.0/#sec-createarrayiterator'
	},
	CreateBuiltinFunction: {
		url: 'https://262.ecma-international.org/6.0/#sec-createbuiltinfunction'
	},
	CreateByteDataBlock: {
		url: 'https://262.ecma-international.org/6.0/#sec-createbytedatablock'
	},
	CreateDataProperty: {
		url: 'https://262.ecma-international.org/6.0/#sec-createdataproperty'
	},
	CreateDataPropertyOrThrow: {
		url: 'https://262.ecma-international.org/6.0/#sec-createdatapropertyorthrow'
	},
	CreateDynamicFunction: {
		url: 'https://262.ecma-international.org/6.0/#sec-createdynamicfunction'
	},
	CreateHTML: {
		url: 'https://262.ecma-international.org/6.0/#sec-createhtml'
	},
	CreateIntrinsics: {
		url: 'https://262.ecma-international.org/6.0/#sec-createintrinsics'
	},
	CreateIterResultObject: {
		url: 'https://262.ecma-international.org/6.0/#sec-createiterresultobject'
	},
	CreateListFromArrayLike: {
		url: 'https://262.ecma-international.org/6.0/#sec-createlistfromarraylike'
	},
	CreateListIterator: {
		url: 'https://262.ecma-international.org/6.0/#sec-createlistiterator'
	},
	CreateMapIterator: {
		url: 'https://262.ecma-international.org/6.0/#sec-createmapiterator'
	},
	CreateMappedArgumentsObject: {
		url: 'https://262.ecma-international.org/6.0/#sec-createmappedargumentsobject'
	},
	CreateMethodProperty: {
		url: 'https://262.ecma-international.org/6.0/#sec-createmethodproperty'
	},
	CreatePerIterationEnvironment: {
		url: 'https://262.ecma-international.org/6.0/#sec-createperiterationenvironment'
	},
	CreateRealm: {
		url: 'https://262.ecma-international.org/6.0/#sec-createrealm'
	},
	CreateSetIterator: {
		url: 'https://262.ecma-international.org/6.0/#sec-createsetiterator'
	},
	CreateUnmappedArgumentsObject: {
		url: 'https://262.ecma-international.org/6.0/#sec-createunmappedargumentsobject'
	},
	DateFromTime: {
		url: 'https://262.ecma-international.org/6.0/#sec-date-number'
	},
	Day: {
		url: 'https://262.ecma-international.org/6.0/#sec-day-number-and-time-within-day'
	},
	DayFromYear: {
		url: 'https://262.ecma-international.org/6.0/#sec-year-number'
	},
	DaylightSavingTA: {
		url: 'https://262.ecma-international.org/6.0/#sec-daylight-saving-time-adjustment'
	},
	DaysInYear: {
		url: 'https://262.ecma-international.org/6.0/#sec-year-number'
	},
	DayWithinYear: {
		url: 'https://262.ecma-international.org/6.0/#sec-month-number'
	},
	Decode: {
		url: 'https://262.ecma-international.org/6.0/#sec-decode'
	},
	DefinePropertyOrThrow: {
		url: 'https://262.ecma-international.org/6.0/#sec-definepropertyorthrow'
	},
	DeletePropertyOrThrow: {
		url: 'https://262.ecma-international.org/6.0/#sec-deletepropertyorthrow'
	},
	DetachArrayBuffer: {
		url: 'https://262.ecma-international.org/6.0/#sec-detacharraybuffer'
	},
	Encode: {
		url: 'https://262.ecma-international.org/6.0/#sec-encode'
	},
	EnqueueJob: {
		url: 'https://262.ecma-international.org/6.0/#sec-enqueuejob'
	},
	EnumerableOwnNames: {
		url: 'https://262.ecma-international.org/6.0/#sec-enumerableownnames'
	},
	EscapeRegExpPattern: {
		url: 'https://262.ecma-international.org/6.0/#sec-escaperegexppattern'
	},
	EvalDeclarationInstantiation: {
		url: 'https://262.ecma-international.org/6.0/#sec-evaldeclarationinstantiation'
	},
	EvaluateCall: {
		url: 'https://262.ecma-international.org/6.0/#sec-evaluatecall'
	},
	EvaluateDirectCall: {
		url: 'https://262.ecma-international.org/6.0/#sec-evaluatedirectcall'
	},
	EvaluateNew: {
		url: 'https://262.ecma-international.org/6.0/#sec-evaluatenew'
	},
	floor: {
		url: 'https://262.ecma-international.org/6.0/#sec-algorithm-conventions'
	},
	ForBodyEvaluation: {
		url: 'https://262.ecma-international.org/6.0/#sec-forbodyevaluation'
	},
	'ForIn/OfBodyEvaluation': {
		url: 'https://262.ecma-international.org/6.0/#sec-runtime-semantics-forin-div-ofbodyevaluation-lhs-stmt-iterator-lhskind-labelset'
	},
	'ForIn/OfHeadEvaluation': {
		url: 'https://262.ecma-international.org/6.0/#sec-runtime-semantics-forin-div-ofheadevaluation-tdznames-expr-iterationkind'
	},
	FromPropertyDescriptor: {
		url: 'https://262.ecma-international.org/6.0/#sec-frompropertydescriptor'
	},
	FulfillPromise: {
		url: 'https://262.ecma-international.org/6.0/#sec-fulfillpromise'
	},
	FunctionAllocate: {
		url: 'https://262.ecma-international.org/6.0/#sec-functionallocate'
	},
	FunctionCreate: {
		url: 'https://262.ecma-international.org/6.0/#sec-functioncreate'
	},
	FunctionInitialize: {
		url: 'https://262.ecma-international.org/6.0/#sec-functioninitialize'
	},
	GeneratorFunctionCreate: {
		url: 'https://262.ecma-international.org/6.0/#sec-generatorfunctioncreate'
	},
	GeneratorResume: {
		url: 'https://262.ecma-international.org/6.0/#sec-generatorresume'
	},
	GeneratorResumeAbrupt: {
		url: 'https://262.ecma-international.org/6.0/#sec-generatorresumeabrupt'
	},
	GeneratorStart: {
		url: 'https://262.ecma-international.org/6.0/#sec-generatorstart'
	},
	GeneratorValidate: {
		url: 'https://262.ecma-international.org/6.0/#sec-generatorvalidate'
	},
	GeneratorYield: {
		url: 'https://262.ecma-international.org/6.0/#sec-generatoryield'
	},
	Get: {
		url: 'https://262.ecma-international.org/6.0/#sec-get-o-p'
	},
	GetBase: {
		url: 'https://262.ecma-international.org/6.0/#sec-jobs-and-job-queues'
	},
	GetFunctionRealm: {
		url: 'https://262.ecma-international.org/6.0/#sec-getfunctionrealm'
	},
	GetGlobalObject: {
		url: 'https://262.ecma-international.org/6.0/#sec-getglobalobject'
	},
	GetIdentifierReference: {
		url: 'https://262.ecma-international.org/6.0/#sec-getidentifierreference'
	},
	GetIterator: {
		url: 'https://262.ecma-international.org/6.0/#sec-getiterator'
	},
	GetMethod: {
		url: 'https://262.ecma-international.org/6.0/#sec-getmethod'
	},
	GetModuleNamespace: {
		url: 'https://262.ecma-international.org/6.0/#sec-getmodulenamespace'
	},
	GetNewTarget: {
		url: 'https://262.ecma-international.org/6.0/#sec-getnewtarget'
	},
	GetOwnPropertyKeys: {
		url: 'https://262.ecma-international.org/6.0/#sec-getownpropertykeys'
	},
	GetPrototypeFromConstructor: {
		url: 'https://262.ecma-international.org/6.0/#sec-getprototypefromconstructor'
	},
	GetReferencedName: {
		url: 'https://262.ecma-international.org/6.0/#sec-jobs-and-job-queues'
	},
	GetSubstitution: {
		url: 'https://262.ecma-international.org/6.0/#sec-getsubstitution'
	},
	GetSuperConstructor: {
		url: 'https://262.ecma-international.org/6.0/#sec-getsuperconstructor'
	},
	GetTemplateObject: {
		url: 'https://262.ecma-international.org/6.0/#sec-gettemplateobject'
	},
	GetThisEnvironment: {
		url: 'https://262.ecma-international.org/6.0/#sec-getthisenvironment'
	},
	GetThisValue: {
		url: 'https://262.ecma-international.org/6.0/#sec-getthisvalue'
	},
	GetV: {
		url: 'https://262.ecma-international.org/6.0/#sec-getv'
	},
	GetValue: {
		url: 'https://262.ecma-international.org/6.0/#sec-getvalue'
	},
	GetValueFromBuffer: {
		url: 'https://262.ecma-international.org/6.0/#sec-getvaluefrombuffer'
	},
	GetViewValue: {
		url: 'https://262.ecma-international.org/6.0/#sec-getviewvalue'
	},
	HasOwnProperty: {
		url: 'https://262.ecma-international.org/6.0/#sec-hasownproperty'
	},
	HasPrimitiveBase: {
		url: 'https://262.ecma-international.org/6.0/#sec-jobs-and-job-queues'
	},
	HasProperty: {
		url: 'https://262.ecma-international.org/6.0/#sec-hasproperty'
	},
	HourFromTime: {
		url: 'https://262.ecma-international.org/6.0/#sec-hours-minutes-second-and-milliseconds'
	},
	ImportedLocalNames: {
		url: 'https://262.ecma-international.org/6.0/#sec-importedlocalnames'
	},
	InitializeHostDefinedRealm: {
		url: 'https://262.ecma-international.org/6.0/#sec-initializehostdefinedrealm'
	},
	InitializeReferencedBinding: {
		url: 'https://262.ecma-international.org/6.0/#sec-initializereferencedbinding'
	},
	InLeapYear: {
		url: 'https://262.ecma-international.org/6.0/#sec-year-number'
	},
	InstanceofOperator: {
		url: 'https://262.ecma-international.org/6.0/#sec-instanceofoperator'
	},
	IntegerIndexedElementGet: {
		url: 'https://262.ecma-international.org/6.0/#sec-integerindexedelementget'
	},
	IntegerIndexedElementSet: {
		url: 'https://262.ecma-international.org/6.0/#sec-integerindexedelementset'
	},
	IntegerIndexedObjectCreate: {
		url: 'https://262.ecma-international.org/6.0/#sec-integerindexedobjectcreate'
	},
	InternalizeJSONProperty: {
		url: 'https://262.ecma-international.org/6.0/#sec-internalizejsonproperty'
	},
	Invoke: {
		url: 'https://262.ecma-international.org/6.0/#sec-invoke'
	},
	IsAccessorDescriptor: {
		url: 'https://262.ecma-international.org/6.0/#sec-isaccessordescriptor'
	},
	IsAnonymousFunctionDefinition: {
		url: 'https://262.ecma-international.org/6.0/#sec-isanonymousfunctiondefinition'
	},
	IsArray: {
		url: 'https://262.ecma-international.org/6.0/#sec-isarray'
	},
	IsCallable: {
		url: 'https://262.ecma-international.org/6.0/#sec-iscallable'
	},
	IsCompatiblePropertyDescriptor: {
		url: 'https://262.ecma-international.org/6.0/#sec-iscompatiblepropertydescriptor'
	},
	IsConcatSpreadable: {
		url: 'https://262.ecma-international.org/6.0/#sec-isconcatspreadable'
	},
	IsConstructor: {
		url: 'https://262.ecma-international.org/6.0/#sec-isconstructor'
	},
	IsDataDescriptor: {
		url: 'ht
